COMPRESS-602 - Migrate zip package to use NIO

PostelnicuGeorge opened the pull request:
Implementation of: https://issues.apache.org/jira/browse/COMPRESS-602
This is a continuation of the story to Use java.nio.file.Path instead of java.io.File throughout commons-compress project.
I would like to highlight that these 2 constructors to be marked as deprecated or prepared for removal because we cannot know what the FileSystem of the caller is and we could only create in the default one.

```java
    public ZipFile(final String name) throws IOException {
        this(new File(name).toPath(), ZipEncodingHelper.UTF8);
    }


    public ZipFile(final String name, final String encoding) throws IOException {
        this(new File(name).toPath(), encoding, true);
    }
```

garydgregory commented on src/main/java/org/apache/commons/compress/archivers/zip/ZipSplitOutputStream.java:
```diff
@@ -57,19 +58,31 @@
+     *
+     * @param zipFile   the zip file to write to
+     * @param splitSize the split size
+     */
```
Missing since tag on new or protected method.

PostelnicuGeorge replied:
@garydgregory I don't understand your comment, can you elaborate cause you mentioned this kind of change multiple times and I don't want to create more problems.

PeterAlfredLee replied:
Hi @PostelnicuGeorge 
@garydgregory is talking about the `@since` tag. You can also find this in comments of other methods.
1.21 is already released so you should add a `@since 1.22` here.

PostelnicuGeorge replied:
Thank you so much, on it!

garydgregory commented on src/main/java/org/apache/commons/compress/archivers/zip/ScatterZipOutputStream.java:
```diff
@@ -198,7 +210,19 @@ public static ScatterZipOutputStream fileBased(final File file) throws FileNotFo
+     * @param compressionLevel The compression level to use, @see #Deflater
+     * @return A  ScatterZipOutputStream that is ready for use.
+     * @throws FileNotFoundException if the path cannot be found
+     */
```
Missing since tag on new or protected method.

garydgregory commented on src/main/java/org/apache/commons/compress/archivers/zip/ZipArchiveOutputStream.java:
```diff
@@ -518,6 +543,11 @@ public void setUseZip64(final Zip64Mode mode) {
         zip64Mode = mode;
     }
 
+    @Override
```
Missing since tag on new or protected method.

garydgregory commented on src/main/java/org/apache/commons/compress/archivers/zip/ZipArchiveOutputStream.java:
```diff
@@ -365,8 +365,33 @@ public ZipArchiveOutputStream(final Path file, final OpenOption... options) thro
+     * @param zipSplitSize maximum size of a single part of the split
+     * archive created by this stream. Must be between 64kB and about
+     * 4GB.
+     *
```
No need for extra whitespace.

PostelnicuGeorge replied:
This is a copy paste from the existing javadocs of methods from within the class, if you want I can correct them all for white-spaces.

garydgregory replied:
I prefer a focused PR, new lines should be "correct". If you want to update unrelated lines, feel free to do so in a _different_ PR, please. This makes reviewing simpler and the PR itself only about one item.

PostelnicuGeorge replied:
Tried my best to undo unrelated changes and keep new things compact and correct.

garydgregory commented on src/main/java/org/apache/commons/compress/archivers/zip/ZipFile.java:
```diff
@@ -174,6 +175,17 @@ public ZipFile(final File f) throws IOException {
+     *
+     * @throws IOException if an error occurs while reading the file.
+     */
+    public ZipFile(final Path path) throws IOException {
```
Missing since tag on new or protected method.

garydgregory commented on src/main/java/org/apache/commons/compress/archivers/zip/ZipFile.java:
```diff
@@ -210,7 +222,21 @@ public ZipFile(final String name, final String encoding) throws IOException {
+     *
+     * @param path path to the archive.
+     * @param encoding the encoding to use for file names, use null
+     * for the platform's default encoding
```
No need for extra whitespace.

PostelnicuGeorge replied:
This is a copy paste from the existing javadocs of methods from within the class, if you want I can correct them all for white-spaces.

garydgregory commented on src/main/java/org/apache/commons/compress/archivers/zip/ZipFile.java:
```diff
@@ -210,7 +222,21 @@ public ZipFile(final String name, final String encoding) throws IOException {
+     * for the platform's default encoding
+     *
+     * @throws IOException if an error occurs while reading the file.
+     */
```
Missing since tag on new or protected method.

garydgregory commented on src/main/java/org/apache/commons/compress/archivers/zip/ZipFile.java:
```diff
@@ -227,7 +253,24 @@ public ZipFile(final File f, final String encoding) throws IOException {
+     * @param encoding the encoding to use for file names, use null
+     * for the platform's default encoding
+     * @param useUnicodeExtraFields whether to use InfoZIP Unicode
+     * Extra Fields (if present) to set the file names.
```
Ditto

garydgregory commented on src/main/java/org/apache/commons/compress/archivers/zip/ZipFile.java:
```diff
@@ -262,6 +305,39 @@ public ZipFile(final File f, final String encoding, final boolean useUnicodeExtr
+     * Opens the given path for reading, assuming the specified
+     * encoding for file names.
+     *
+     *
```
Ditto

PostelnicuGeorge replied:
This is a copy paste from the existing javadocs of methods from within the class, if you want I can correct them all for white-spaces.

garydgregory commented on src/main/java/org/apache/commons/compress/archivers/zip/ZipSplitReadOnlySeekableByteChannel.java:
```diff
@@ -47,7 +50,7 @@
     private static final int ZIP_SPLIT_SIGNATURE_LENGTH = 4;
     private final ByteBuffer zipSplitSignatureByteBuffer =
-        ByteBuffer.allocate(ZIP_SPLIT_SIGNATURE_LENGTH);
+            ByteBuffer.allocate(ZIP_SPLIT_SIGNATURE_LENGTH);
```
Please do not change the formatting, it makes reviewing harder and take more time.

PostelnicuGeorge replied:
My mistake, reverted!

garydgregory commented on src/main/java/org/apache/commons/compress/archivers/zip/ZipSplitReadOnlySeekableByteChannel.java:
```diff
@@ -58,11 +61,11 @@
-     * @throws IOException if the first channel doesn't seem to hold
-     * the beginning of a split archive
+     * @throws IOException          if the first channel doesn't seem to hold
+     *                              the beginning of a split archive
```
Ditto

PostelnicuGeorge replied:
My mistake, reverted!

garydgregory commented on src/main/java/org/apache/commons/compress/archivers/zip/ZipSplitReadOnlySeekableByteChannel.java:
```diff
@@ -58,11 +61,11 @@
      */
     public ZipSplitReadOnlySeekableByteChannel(final List<SeekableByteChannel> channels)
-        throws IOException {
+            throws IOException {
```
Don't change formatting.

PostelnicuGeorge replied:
My mistake, reverted!

garydgregory commented on src/main/java/org/apache/commons/compress/archivers/zip/ZipSplitReadOnlySeekableByteChannel.java:
```diff
@@ -108,10 +111,10 @@ private void assertSplitSignature(final List<SeekableByteChannel> channels)
      * @return SeekableByteChannel that concatenates all provided channels
      * @throws NullPointerException if channels is null
-     * @throws IOException if reading channels fails
+     * @throws IOException          if reading channels fails
```
Ditto

PostelnicuGeorge replied:
My mistake, reverted!

garydgregory commented on src/main/java/org/apache/commons/compress/archivers/zip/ZipSplitReadOnlySeekableByteChannel.java:
```diff
@@ -125,14 +128,14 @@ public static SeekableByteChannel forOrderedSeekableByteChannels(final SeekableB
      */
     public static SeekableByteChannel forOrderedSeekableByteChannels(final SeekableByteChannel lastSegmentChannel,
-        final Iterable<SeekableByteChannel> channels) throws IOException {
+                                                                     final Iterable<SeekableByteChannel> channels) throws IOException {
```
Again, change messes up formatting.

PostelnicuGeorge replied:
My mistake, reverted!

garydgregory commented on src/main/java/org/apache/commons/compress/utils/FileNameUtils.java:
```diff
@@ -50,6 +51,19 @@ public static String getExtension(final String filename) {
         return name.substring(extensionPosition + 1);
     }
 
+    public static String getExtensionFrom(final Path path) {
```
Make this private or package private. The fewer public or protected method the better as they will have to be maintained "forever". Also needs a comment.

garydgregory commented on src/main/java/org/apache/commons/compress/utils/FileNameUtils.java:
```diff
@@ -75,4 +89,19 @@ public static String getBaseName(final String filename) {
         return name.substring(0, extensionPosition);
     }
+
+    public static String getBaseNameFrom(final Path path) {
```
As above plus needs a comment.

PostelnicuGeorge replied:
Added javadocs

garydgregory commented on src/test/java/org/apache/commons/compress/archivers/zip/ScatterZipOutputStreamTest.java:
```diff
@@ -45,7 +46,8 @@ public void cleanup() {
-        final ScatterZipOutputStream scatterZipOutputStream = ScatterZipOutputStream.fileBased(scatterFile);
+        final ScatterZipOutputStream scatterZipOutputStream = ScatterZipOutputStream.fileBased(scatterFile,
+                Deflater.BEST_COMPRESSION);
         final byte[] B_PAYLOAD = "RBBBBBBS".getBytes();
```
Hm, why? How is this part of the NIO change? Hard to te with so many changes...

PostelnicuGeorge replied:
Wanted to increase the coverage and test this branch as well.

garydgregory replied:
Please keep the PR focused on one thing, this is a behavior change unrelated to NIO. May you revert this line please?

PostelnicuGeorge replied:
Reverted.

garydgregory commented on src/test/java/org/apache/commons/compress/archivers/zip/ZipSplitOutputStreamTest.java:
```diff
@@ -69,19 +69,19 @@ public void splitZipBeginsWithZipSplitSignature() throws IOException {
         final File testOutputFile = new File(dir, "testCreateSplittedFiles.zip");
         final int splitSize = 100 * 1024; /* 100KB */
-        final ZipSplitOutputStream zipSplitOutputStream = new ZipSplitOutputStream(testOutputFile, splitSize);
+        final ZipSplitOutputStream ZipSplitOutputStream = new ZipSplitOutputStream(testOutputFile, splitSize);
```
-1: Local var names never start with a CAPITAL letter.

PostelnicuGeorge replied:
My mistake, don't know how this happened or why I modified this test :/

garydgregory commented on src/test/java/org/apache/commons/compress/archivers/zip/ZipSplitOutputStreamTest.java:
```diff
@@ -69,19 +69,19 @@ public void splitZipBeginsWithZipSplitSignature() throws IOException {
 
         while ((readLen = inputStream.read(buffer)) > 0) {
-            zipSplitOutputStream.write(buffer, 0, readLen);
+            ZipSplitOutputStream.write(buffer, 0, readLen);
```
-1 ditto

PostelnicuGeorge replied:
My mistake, don't know how this happened or why I modified this test :/

garydgregory reviewed (changes requested):
Thank you for your PR. That's a lot of changes. This needs a more thorough review after you address my comments. Do not change the formatting; this just creates annoying changes that still take time to check. I hopefully can review again over the weekend.

PeterAlfredLee commented on src/main/java/org/apache/commons/compress/archivers/zip/ParallelScatterZipCreator.java:
```diff
@@ -17,13 +17,11 @@
-import org.apache.commons.compress.parallel.InputStreamSupplier;
-import org.apache.commons.compress.parallel.ScatterGatherBackingStore;
-import org.apache.commons.compress.parallel.ScatterGatherBackingStoreSupplier;
+import org.apache.commons.compress.parallel.*;
```
We do not use wildcard imports. Please specify the packages you want to import.

garydgregory replied:
> We do not use wildcard imports. Please specify the packages you want to import.

Good catch, I think Checkstyle can catch that.

PostelnicuGeorge replied:
I have undone all .* imports in the whole zip package

PostelnicuGeorge commented:
Added since tags to all the new public methods I have added, hopefully I haven't missed something, I have also removed all .* star imports and removed all the white-spaces added by my formatting.

garydgregory commented:
Hi @PostelnicuGeorge 

Thank you for your updates.

When I look at the PR as a [diff](https://patch-diff.githubusercontent.com/raw/apache/commons-compress/pull/236.diff) I still see missing Javadoc since tags. 

Let's keep this PR focused on just NIO without other changes.

Please see my other comments.

TY!

PostelnicuGeorge commented:
Hi @garydgregory ,

Tried my best to revert all changes that are unrelated to NIO, keep the javadocs compact and correct, code is updated with latest master as well.
Let me know if I missed other things.

garydgregory commented on src/main/java/org/apache/commons/compress/parallel/FileBasedScatterGatherBackingStore.java:
```diff
@@ -23,21 +23,26 @@
+        this(target.toPath());
+    }
+
+    public FileBasedScatterGatherBackingStore(final Path target) throws FileNotFoundException {
```
This new eleemnt needs a Javadoc comment and since tag.

PostelnicuGeorge replied:
I changed in place the javadoc at class level, do you want me to put it at the constructor level?
```java
/**
 * ScatterGatherBackingStore that is backed by a file or path.
 *
 * @since 1.22
 */
```

PostelnicuGeorge replied:
Reverted change to class level javadoc, added a new one at constructor level with since tag.

garydgregory reviewed (changes requested):
@PostelnicuGeorge 
This PR looks really much better, TY! I found one last new element that needs a Javadoc comment and since tag. See other comment.

kinow commented on src/main/java/org/apache/commons/compress/archivers/zip/ZipSplitReadOnlySeekableByteChannel.java:
```diff
@@ -155,30 +158,40 @@ public static SeekableByteChannel forOrderedSeekableByteChannels(final SeekableB
+    }
+
+    /**
+     * Concatenates zip split files from the last segment(the extension SHOULD be .zip)
```
Missing space between segment and parenthesis, and would SHOULD be replaced by MUST? It repeats in other javadocs further down in this file :point_down:

PostelnicuGeorge replied:
hey, I changed in the places were I changed, but won't change in the places that were before, like Gary said to keep modifications only on the scope of the task.

kinow commented on src/main/java/org/apache/commons/compress/utils/FileNameUtils.java:
```diff
@@ -50,6 +51,29 @@ public static String getExtension(final String filename) {
+
+        final String name = path.getFileName().toString();
+        final int extensionPosition = name.lastIndexOf('.');
+        if (extensionPosition < 0) {
```
Was going to comment on the public vs package/private here too, but I see @garydgregory already asked about it :point_up: 

@PostelnicuGeorge I think we can change here to 

```suggestion
        if (extensionPosition < 0 || extensionPosition == name.length() - 1) {
```

? This way we avoid calling `.substring` when the strings ends with a dot.

PostelnicuGeorge replied:
The package private can't be because of the usages and due to the fact that I would like that these 2 methods will replace the original ones in the tests and in the main code usages. Also I didn't try to change the functionality that was there before only to make it work for nio, that's why I didn't attempt any enhancements anywhere only equivalent methods from io to nio.

kinow reviewed (commented):
Was interested to see what was changing, so did a quick review :+1: 

Thanks @PostelnicuGeorge !

garydgregory reviewed (approved):
Hi @PostelnicuGeorge 
I updated and pushed this GitHub PR with all of my comments
from today addressed which was simpler than going round and round and
re-reviewing a large PR. 
TY!

Credit is in changes.xml.

PostelnicuGeorge commented on src/main/java/org/apache/commons/compress/parallel/FileBasedScatterGatherBackingStore.java:
```diff
@@ -70,8 +80,8 @@ public void close() throws IOException {
-            if (target.exists() && !target.delete()) {
-                target.deleteOnExit();
+            if (Files.exists(target)) {
+                Files.deleteIfExists(target);
```
Hey Gary you're right, I've read about StandardOpenOption.DELETE_ON_CLOSE and it only works in the same try-with-resources, the way we're passing streams around doesn't so I think we need to think about a general solution or change the way we're using different things.

PostelnicuGeorge commented on src/main/java/org/apache/commons/compress/utils/FileNameUtils.java:
```diff
@@ -50,6 +51,29 @@ public static String getExtension(final String filename) {
+     * @param path the path of the file to obtain the extension of.
+     * @since 1.22
+     */
+    public static String getExtensionFrom(final Path path) {
```
I thought of them as utility methods and it didn't even occur to me to unit test them. Will do so in the future!
